rust/rsc/src/rsc/main.rs

```diff
@@ -213,7 +217,55 @@ fn launch_eviction(conn: Arc<DatabaseConnection>, tick_interval: u64, deadline:
                 .await
                 .unwrap();
 
-            tracing::info!(%res.rows_affected, "Performed TTL eviction tick.");
+            tracing::info!(%res.rows_affected, "Performed Job TTL eviction tick.");
+        }
+    });
+}
+
+fn launch_blob_eviction(
+    conn: Arc<DatabaseConnection>,
+    tick_interval: i64,
+    deadline: i64,
+    blob_stores: HashMap<Uuid, Arc<dyn blob::DebugBlobStore + Sync + Send>>,
+) {
+    // TODO: This should probably be a transaction so that a job can't add a new reference to a
+    // blob as we are deleting it.
+
+    tokio::spawn(async move {
+        let mut interval = tokio::time::interval(std::time::Duration::from_secs(
+            tick_interval.try_into().unwrap(),
+        ));
+        loop {
+            interval.tick().await;
+
+            // Blobs must be at least this old to be considered for eviction.
+            // This gives clients time to reference a blob before it gets evicted.
+            let deadline = (Utc::now() - Duration::seconds(deadline)).naive_utc();
+
+            let blobs = blob_store_service::fetch_unreferenced_blobs(conn.as_ref(), deadline)
+                .await
+                .unwrap();
+            let blob_ids: Vec<Uuid> = blobs.iter().map(|blob| blob.id).collect();
+            let mut affected: u64 = blob_ids.len().try_into().unwrap();
+
+            if affected == 0 {
+                tracing::info!(%affected, "Performed Blob TTL eviction tick.");
+                continue;
+            }
+
+            // Delete blobs from database
+            let res = blob_store_service::delete_blobs_by_ids(conn.as_ref(), blob_ids)
+                .await
+                .unwrap();
+            affected = res.rows_affected;
+
+            // Delete blobs from blob store
+            for blob in blobs {
+                let store = blob_stores.get(&blob.store_id).unwrap();
+                store.delete_key(blob.key).await.unwrap();
```

Comment:
What happens if the database removal succeeds but the store removal fails? Do we need some recovery mechanism? or should we have some kind of critical block here to ensure both things always happen?

Reply:
This _will_ happen and there is pretty much nothing we can do about it here.

The store removal can be a remove server taking 100s of times longer to respond than we are willing to wait. (in fact, I should probably spawn a thread for the call to delete_key). 

This outcome is better than the reverse imo. We get to choose from
1) exists in the database but not the blob store
2) exists in the blob store but not the database

One thing we can do in the long term to mitigate #2 is to have a periodic task for "go collect all blobs on a store, figure out which ones aren't referenced, try to delete them again, make a report for human review". #1 doesn't have as much headache in later cleanup but it means that there is a much higher chance that a client will be given a reference to a blob that doesn't exist (which to be fair, we have to be tolerant of this regardless)